capirca/lib/versa.py

```diff
@@ -0,0 +1,877 @@
+  pass
+
+class VersaVerbatimError(Error):
+  """generic error class."""
+  pass
+
+
+class VersaOptionError(Error):
+  """generic error class."""
+  pass
+
+
+class MixedAddrBookTypesError(Error):
+  """generic error class."""
+  pass
+
+
+class ConflictingTargetOptionsError(Error):
+  """generic error class."""
+  pass
+
+
+class ConflictingApplicationSetsError(Error):
+  """generic error class."""
+  pass
+
+class Tree:
+  """Creates a Tree Object."""
+  target=[]
+  INDENT = '    '
+
+  def __init__(self, name='root',typ=None):
+    """The init function"""
+    self.children = []
+    self.name = name
+    self.typ = typ
+
+
+  def AddParent(self, parent=None):
+    """Add a Node to Parent"""
+    if isinstance(parent, Tree):
+      parent.AddNode(self)
+
+  def __repr__(self):
+    """repr for a Node """
+    return self.name
+
+  def AddNode(self, node):
+    """add a Node """
+    assert isinstance(node, Tree)
+    self.children.append(node)
+
+  def FindNode(self, nodename):
+    """find a Node """
+    if self.name == nodename:
+      return self
+    if self.children:
+      for child in self.children:
+        ret = child.FindNode(nodename)
+        if isinstance(ret, Tree):
+          return ret
+    return None
+
+  # Print the tree
+  def PrintTree(self,num=0):
+    """Print to the tree.It returns the target """
+    self.ResetTarget()
+    self.PrintTreeInt(num)
+    return self.target
+
+  def PrintTreeInt(self,num=0):
+    """Internal function to the tree. Does recursion"""
+    if len(self.name) > 0:
+      self.target.append(f'{self.INDENT*num}{self.name}' + '  {')
+    if self.typ is not None:
+      if isinstance(self.typ, str):
+        if len(self.name) > 0:
+          self.target.append(f'{self.INDENT*(num+1)}{self.typ}')
+        else:
+          self.target.append(f'{self.INDENT*num}{self.typ}')
+      elif isinstance(self.typ, list):
+        for item in self.typ:
+          self.target.append(f'{self.INDENT*(num+1)}{item}')
+    if self.children:
+      for child in self.children:
+        child.PrintTreeInt(num+1)
+    if len(self.name) > 0:
+      self.target.append(f'{self.INDENT*num}' + '}')
+
+  def ResetTarget(self):
+    """Reset the target """
+    self.target.clear()
+
+class Term(aclgenerator.Term):
+  """Representation of an individual Versa term.
+
+     This is mostly useful for the __str__() method.
+
+     Args:
+       obj: a policy.Term object
+       filter_options: list of remaining target options (zones)
+  """
+
+  ACTIONS = {'accept': 'permit',
+             'deny': 'deny',
+             'reject': 'reject',
+             'count': 'count',
+             'log': 'log',
+             'dscp': 'dscp'}
+
+  def __init__(self, term, from_zone, to_zone, addrbook, verbose=True):
+    super().__init__(term)
+    self.term = term
+    self.from_zone = []
+    self.to_zone = []
+    self.verbose = verbose
+    self.addrbook = addrbook
+    self.app = []
+
+    if term.source_zone:
```

Comment:
qq on Versa: Is intra-Zone traffic permitted by default, or denied by default?

Reply:
Deny